third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -236,6 +247,123 @@ impl<'a> FunctionGenerator<'a> {
         }
     }
 
+    /// Remove pairs of instructions that are equivalent to a no-op.
+    fn remove_redundant_instr_pairs(&mut self) {
+        let mut changed = true;
+        let mut deleted = vec![false; self.code.len()];
+        let mut no_deletions = true; // same as `deleted.iter().all(|v| !*v)`
+        let mut is_jump_target = vec![false; self.code.len()];
+        for info in self.label_info.values() {
+            if let Some(label_offs) = info.resolution {
+                if !info.references.is_empty() {
+                    is_jump_target[label_offs as usize] = true;
+                }
+            }
+        }
+        // Note: this is a worst case n^2 algorithm, but it is expected to be linear in practice.
+        // It marks instructions to be deleted lazily.
+        // If this becomes a bottleneck, we can implement a worst case linear algorithm (by
+        // maintaining a doubly-linked list of next-previous undeleted instruction indices).
+        while changed {
+            // repeat until no more deletions
+            changed = false;
+            let mut this_index = 0;
+            while this_index < (self.code.len() - 1) {
+                if deleted[this_index] {
+                    this_index += 1;
+                    continue;
+                }
+                let this_bc = &self.code[this_index];
+                if let Some(mut next_index) = deleted[(this_index + 1)..].iter().position(|v| !*v) {
+                    next_index += this_index + 1;
+                    // Note that we should not delete a sequence of pairs of instructions if we can jump to
+                    // any of them (except the first one).
+                    if is_jump_target[(this_index + 1)..=next_index]
+                        .iter()
+                        .all(|v| !*v)
+                    {
+                        let next_bc = &self.code[next_index];
+                        use FF::Bytecode::*;
+                        // Each of these pairs of instructions are equivalent to a no-op, so they can be safely
+                        // removed assuming the above conditions are met.
+                        match (this_bc, next_bc) {
```

Comment:
Perhaps in the first iteration we do not need the traits but at least we can break this function down in a more compositional way instead of hardcoding the modification logic inside of this loop

Reply:
Done, implemented traits.